src/compiler/emitter.ts

```diff
@@ -4256,75 +4267,99 @@ namespace ts {
-                    return !rangeIsOnSingleLine(parentNode, currentSourceFile!);
+                    return rangeIsOnSingleLine(parentNode, currentSourceFile!) ? 0 : 1;
                 }
-                else if (positionIsSynthesized(parentNode.pos) || nodeIsSynthesized(firstChild)) {
-                    return synthesizedNodeStartsOnNewLine(firstChild, format);
+                else if (!positionIsSynthesized(parentNode.pos) && !nodeIsSynthesized(firstChild) && firstChild.parent === parentNode) {
+                    if (preserveNewlines) {
+                        const lines = getLinesBetweenPositionAndPrecedingNonWhitespaceCharacter(firstChild.pos, currentSourceFile!, /*includeComments*/ true);
+                        return lines === 0
+                            ? getLinesBetweenPositionAndPrecedingNonWhitespaceCharacter(firstChild.pos, currentSourceFile!, /*includeComments*/ false)
+                            : lines;
+                    }
+                    return rangeStartPositionsAreOnSameLine(parentNode, firstChild, currentSourceFile!) ? 0 : 1;
                 }
-                else {
-                    return !rangeStartPositionsAreOnSameLine(parentNode, firstChild, currentSourceFile!);
+                else if (synthesizedNodeStartsOnNewLine(firstChild, format)) {
+                    return 1;
                 }
             }
-            else {
-                return false;
-            }
+            return format & ListFormat.MultiLine ? 1 : 0;
         }
 
-        function shouldWriteSeparatingLineTerminator(previousNode: Node | undefined, nextNode: Node, format: ListFormat) {
-            if (format & ListFormat.MultiLine) {
-                return true;
-            }
-            else if (format & ListFormat.PreserveLines) {
+        function getSeparatingLineTerminatorCount(previousNode: Node | undefined, nextNode: Node, format: ListFormat): number {
+            if (format & ListFormat.PreserveLines || preserveNewlines) {
                 if (previousNode === undefined || nextNode === undefined) {
-                    return false;
+                    return 0;
                 }
-                else if (nodeIsSynthesized(previousNode) || nodeIsSynthesized(nextNode)) {
-                    return synthesizedNodeStartsOnNewLine(previousNode, format) || synthesizedNodeStartsOnNewLine(nextNode, format);
+                else if (!nodeIsSynthesized(previousNode) && !nodeIsSynthesized(nextNode) && previousNode.parent === nextNode.parent) {
+                    const lines = getEffectiveLinesBetweenRanges(previousNode, nextNode);
+                    return preserveNewlines ? lines : Math.min(lines, 1);
                 }
-                else {
-                    return !rangeEndIsOnSameLineAsRangeStart(previousNode, nextNode, currentSourceFile!);
+                else if (synthesizedNodeStartsOnNewLine(previousNode, format) || synthesizedNodeStartsOnNewLine(nextNode, format)) {

```

Comment:
does `synthesizedNodeStartsOnNewline` require its parameter to synthetic? It might not be in the case that `previousNode.parent !== nextNode.parent`.

Reply:
It short circuits to considering `format` alone if the node isn’t synthesized. I guess it works out because all the tests pass (and it took forever to get them all to pass—when virtually every compiler baseline runs through the emitter, coverage is pretty good), but it probably wasn’t my intent to go into this branch in that case. Will revisit.

Reply:
It turns out that this oversight is what produces some of the desirable baseline changes like https://github.com/microsoft/TypeScript/pull/36688/files#diff-ad4da0479d0f311e5380216f1820cba3. It does seem confusing but I’m not sure what the best way is to be more explicit about what should happen without nesting more conditionals and getting these functions more in the weeds 🤔